packages/@aws-cdk/assets/lib/fs/utils.ts

```diff
@@ -58,3 +59,79 @@ export function shouldFollow(mode: FollowMode, sourceRoot: string, realPath: str
     return path.resolve(realPath).startsWith(path.resolve(sourceRoot));
   }
 }
+
+export function listFilesRecursively(dir: string, options: CopyOptions & Required<Pick<CopyOptions, 'follow'>>, rootDir?: string): string[] {
+  const files = [];
+
+  let exclude = [...(options.exclude || [])];
+  rootDir = rootDir || dir;
+  {
+    const stat = fs.statSync(dir);
+    if (!stat) {
+      return [];
+    }
+
+    if (!stat.isDirectory()) {
+      return [dir];
+    }
+  }
+
+  for (const file of fs.readdirSync(dir)) {
+    let fullFilePath = path.join(dir, file);
+
+    let stat: fs.Stats | undefined = options.follow === FollowMode.ALWAYS
+      ? fs.statSync(fullFilePath)
+      : fs.lstatSync(fullFilePath);
+
+    if (!stat) {
+      continue;
+    }
+
+    if (stat.isSymbolicLink()) {
+      const target = fs.readlinkSync(fullFilePath);
+
+      // determine if this is an external link (i.e. the target's absolute path
+      // is outside of the root directory).
+      const targetPath = path.normalize(path.resolve(dir, target));
+
+      if (shouldFollow(options.follow, rootDir, targetPath)) {
+        stat = fs.statSync(fullFilePath);
+        if (!stat) {
+          continue;
+        }
+
+        fullFilePath = target;
+      }
+    }
+
+    const relativeFilePath = path.relative(rootDir, fullFilePath);
+
+    // we've just discovered that we have a directory
+    if (stat.isDirectory()) {
+      // to help future shouldExclude calls, we're changing the exlusion patterns
+      // by expliciting "dir" exclusions to "dir/*" (same with "!dir" -> "!dir/*")
+      exclude = exclude.reduce<string[]>((res, pattern) => {
```

Comment:
I don't think this should be done in-line while traversing. It is also possible to run through the list of entries in the ignore list, see which ones are directories and treat those as prefixes.

Reply:
In fact, I'd like to convert the "list of exclude/include entries" into a function or class that encapsulates the behavior as soon as possible and pass that around.

I'm thinking something like:

```
const ignoreList = parseIgnoreList(rootDirectory, ...list of strings);
```

`ignoreList` will be either of type `(path: string) => boolean`, or be a class with a method of that type.

I would have that as an argument to `listFilesRecursively()`.

Reply:
The issue is that it wouldn't be very economical, `fs` calls-wise. I'm filling the `exclude` directory on the fly because I'm also checking the `stat` of the files I'm traversing.